recirq/quantum_chess/swap_update_transformer.py

```diff
@@ -0,0 +1,180 @@
+
+    def generate_candidate_swaps(
+        self, gates: Iterable[cirq.Operation]
+    ) -> Generator[Tuple[cirq.GridQubit, cirq.GridQubit], None, None]:
+        """Generates the candidate SWAPs that would have a positive effect on at
+        least one of the given physical gates.
+
+        Args:
+          gates: the list of gates to consider which operate on GridQubits
+        """
+        for gate in gates:
+            for gate_q in gate.qubits:
+                yield from (
+                    (gate_q, swap_q)
+                    for swap_q in self._adjacent_qubits(gate_q)
+                    if mcpe.effect_of_swap((gate_q, swap_q), gate.qubits) > 0)
+
+    def update_iteration(self) -> Generator[cirq.Operation, None, None]:
+        """Runs one iteration of the swap update algorithm and updates internal
+        state about the original circuit.
+
+        Returns:
+          the operations on GridQubits in the final updated circuit generated by
+          this iteration
+        """
+        # Handle the already-satisfied active gates.
+        # Those can be immediately added into the final circuit.
+        active_physical_gates = []
+        for gate in self.dlists.active_gates():
+            physical_gate = gate.transform_qubits(self.mapping.physical)
+            if _satisfies_adjacency(physical_gate):
+                # physical_gate is ready to be popped off the dependecy lists
+                # and added to the final circuit.
+                for q in gate.qubits:
+                    self.dlists.pop_front(q)
+                yield physical_gate
+            else:
+                # physical_gate needs to be fixed up with some swaps.
+                active_physical_gates.append(physical_gate)
+
+        # If all the active gates in this pass were added to the final circuit,
+        # then we have nothing left to do until we make another pass and get the
+        # newly active gates.
+        if not active_physical_gates:
+            return
+
+        candidates = set(self.generate_candidate_swaps(active_physical_gates))
+        # This could happen on irregular topologies or when some device
+        # qubits are disallowed.
+        assert candidates, 'no swaps found that will improve the circuit'
+        chosen_swap = max(
+            candidates,
+            key=lambda swap: self.dlists.maximum_consecutive_positive_effect(
+                *swap, self.mapping))
+        self.mapping.swap_physical(*chosen_swap)
+        yield from self.swap_factory(*chosen_swap)
+
+    def add_swaps(self) -> Generator[cirq.Operation, None, None]:
+        """Iterates the swap update algorithm to completion.
+
+        If the updater already completed, does nothing.
+
+        Returns:
+          the generated operations on physical GridQubits in the final circuit
+        """
+        while not self.dlists.all_empty():
+            yield from self.update_iteration()
+
+
+class SwapUpdateTransformer(ct.CircuitTransformer):
```

Comment:
wondering what your thoughts are on how we will combine our changes?

the initial mapping logic i implemented are technically all util functions and can't be used as a standalone transformer since it assumes some swap update algorithm will update its mapping after.

i understand `SwapUpdater` can be used with any arbitrary initial mapping, but perhaps we should treat it as a util class and other transformer implementations can explicitly call it if they choose to? please correct me if i'm wrong, but my understanding of the `CircuitTransformer` abstraction is that it'll take a device and a circuit and automagically return a circuit that will work on the device without the caller having to know to sometimes create an initial mapping first.

i'll move all the logic in `DynamicLookAheadHeuristicCircuitTransformer` (aside from `transform`) to a separate util file in my PR. we could then either rename this class to `DynamicLookAheadHeuristicCircuitTransformer` or just leave it as is, and update the `transform` method to call the (to be refactored) initial mapping utils and `SwapUpdater`? or we could then also move this class to `circuit_transformer.py` since it'll be relatively tiny and have this file contain `SwapUpdater` related logic for other transformers to use?

we can ofc do the refactoring in a subsequent PR but just wanted to make sure we are on the same page!

Reply:
Actually, I was planning to use the SwapUpdater directly in your DynamicLookAheadHeuristicCircuitTransformer implementation; it sounds like we're on the same page there.

Initially I had added this stand-alone transformer because CircuitTransformer sounds like a more generic circuit -> circuit mapper without any other constraints on the inputs or outputs, and who knows, maybe we'll end up with other kinds of initial placements that would benefit from applying the swap updates.
But I changed my mind -- QuantumBoard requires CircuitTransformer output to be valid and runnable, even if that's not explicit in the interface. And even if we do end up with other kinds of initial placements, there's no reason we can't use SwapUpdater directly for those either. Having the transformer doesn't really add much.

Reply:
awesome sg!